GUIDELINES.md

```diff
@@ -0,0 +1,135 @@
+    * In essence, sweetest has an own small dependency injection framework included
+  * A dependency can be either mocked or real
+    * By default in sweetest all dependencies are considered mocks
+    * If you want to put a class e.g. the `LoginViewModel` under test you need to tell sweetest (see below)
+    * By doing so sweetest will create an instance of `LoginViewModel` and on the way examines the constructor
+    * sweetest is trying to create and inject all dependencies `LoginViewModel` needs by doing the same for all of these dependencies (so e.g. if the constructor of `LoginViewModel` looks like `constructor(private val authManager: AuthManager)` then sweetest will try to create `AuthManager`, too)
+    * As `AuthManager` has not previously been configured to be "real" sweetest automatically creates a Mockito mock for it
+    * All instances created are saved in sweetest's dependency management, so any subsequent request of that dependency will receive the same instances (e.g. if a hypothetical `OtherViewModel` would have a dependency `AuthManager` too then it would be provided the exact same instance as for the `LoginViewModel`)
+  * What's the difference between mock and fake?
+    * A **fake** is an implementation that is done for test conditions, so it should basically behave like the real implementation (so e.g. a `BackendGatewayFakeSteps` would contain an actual list of registered users that gets bigger as soon as a new user is registered and is used to verify authentication attempts)
+    * A **mock** doesn't need to represent the functionality of a real object but can be tailored to what's needed in test conditions
+    * Both of them can be used in order to perform verifications
+* How to add a dependency
+  * You need to put each dependency you use in sweetest into the **module testing configuration** (see above)
+    * Add a dependency by adding `dependency any of<LoginViewModel>` to the configuration (you need to do this before you introduce the dependency to any test or steps class!)
+    * Consider mock vs. real
+      * For example, if you are testing just the `LoginViewModel`, you set sweetest up so it creates a _real_ `LoginViewModel` only (if you don't tell sweetest it will always create a mock for that type)
+      * For example, if you are testing the integration of the `LoginViewModel` with `AuthManager` and `SessionStore` then you need to tell sweetest to create real instances for all of these (`BackendGateway` will remain mocked, though)
+    * Set mode constraints in configuration
+      * In your test systems, if you want to have a certain type always to be **real**, then you can tell sweetest in the configuration like this: `dependency realOnly of<LoginViewModel>` (that tells that `LoginViewModel` might not be used as a dependency in a mocked form; for `LoginViewModel` that makes total sense)
```

Comment:
Can we please get rid of the `realOnly` or `mockOnly` configs?
I often had the problem that somebody assumed "that always has to be mock" or "that's always real" and I needed the exact opposite. It's fix such a early mis-decision, it involves changing loads of files...

Reply:
i was working on getting rid of the whole module configuration as such, couldn't find the priority to actually finish it. in the meantime we could make all dependencies `any` in the config and push all further configuration down to the tests and steps by convention (here). what do you think of that?